File: tests/test_yt_player.py
"""Tests for app/static/js/yt_player.js (Day 8: unified player wrapper).

Mirrors tests/test_transcript_follow.py: the actual JS tests live in
test_yt_player.mjs and are run via `node --test`. This wrapper:
  1. Verifies the .mjs suite passes end-to-end (catches accidental
     test regressions that would otherwise silently rot).
  2. Verifies the video.html template includes the script + iframe id
     + YTPlayer.init() bootstrap, so a deploy-time HTML regression
     breaks loudly here rather than at 11pm.

Why two layers:
  - .mjs = functional coverage (storage, idempotency, backend detection)
  - .py = integration glue (template wiring + .mjs subprocess)

Run with: pytest tests/test_yt_player.py -v
"""

from __future__ import annotations

import re
import shutil
import subprocess
from pathlib import Path

import pytest

REPO_ROOT = Path(__file__).resolve().parent.parent
JS_TEST_FILE = REPO_ROOT / "tests" / "test_yt_player.mjs"
JS_SOURCE_FILE = REPO_ROOT / "app" / "static" / "js" / "yt_player.js"
VIDEO_TEMPLATE = REPO_ROOT / "app" / "templates" / "video.html"


def _has_node() -> bool:
    return shutil.which("node") is not None


pytestmark = pytest.mark.skipif(
    not _has_node(), reason="node not available; cannot run JS test suite"
)


def _run_node_tests() -> tuple[int, str]:
    proc = subprocess.run(
        ["node", "--test", str(JS_TEST_FILE)],
        cwd=str(REPO_ROOT),
        capture_output=True,
        text=True,
        timeout=60,
    )
    return proc.returncode, (proc.stdout or "") + (proc.stderr or "")


def test_node_test_suite_exits_zero():
    """The full node --test suite must pass.

    We don't pin the exact test count (Day 8 added 19, more may come
    in Day 9+), but we assert # pass > 0 and the exit code is 0.
    """
    rc, output = _run_node_tests()
    assert rc == 0, (
        f"node --test failed.\n--- output ---\n{output}\n--- end output ---"
    )
    # Sanity: TAP summary line.
    assert "# pass " in output, (
        f"Expected '# pass' in TAP output, got:\n{output}"
    )


def _parse_tap_test_names(output: str) -> list[tuple[bool, str]]:
    results: list[tuple[bool, str]] = []
    for line in output.splitlines():
        m = re.match(r"^(ok|not ok)\s+\d+\s+-\s+(.*)$", line)
        if m:
            passed = m.group(1) == "ok"
            results.append((passed, m.group(2).strip()))
    return results


def test_node_tests_cover_storage_helpers():
    """The storage sanitization + save/load helpers must be tested.

    These are the security boundary against localStorage poisoning;
    regressions here would let attacker-controlled video IDs corrupt
    other videos' resume positions.
    """
    rc, output = _run_node_tests()
    assert rc == 0, output
    names = [name for _, name in _parse_tap_test_names(output)]
    assert any("storageKey" in n for n in names), names
    assert any("savePosition" in n for n in names), names
    assert any("loadPosition" in n for n in names), names
    assert any("disabled localStorage" in n for n in names), names


def test_node_tests_cover_backend_detection():
    """Backend detection (iframe vs <video> vs neither) must be tested.

    This is the foundation of the whole wrapper — get it wrong and
    every other feature is broken.
    """
    rc, output = _run_node_tests()
    assert rc == 0, output
    names = [name for _, name in _parse_tap_test_names(output)]
    assert any("iframe" in n for n in names), names
    assert any("<video>" in n for n in names), names
    assert any("null backend" in n for n in names), names


def test_node_tests_cover_source_level_regressions():
    """Source-level guards must be tested (canary against regressions).

    - No querySelector('video') outside detectBackend()
    - Native backend doesn't load the YouTube iframe_api
    - Public API surface is stable
    """
    rc, output = _run_node_tests()
    assert rc == 0, output
    names = [name for _, name in _parse_tap_test_names(output)]
    assert any("querySelector" in n for n in names), names
    assert any("iframe_api" in n for n in names), names
    assert any("public API" in n for n in names), names


def test_node_tests_cover_enable_resume():
    """The auto-resume feature must be tested.

    Critical for the Day 8 UX promise (user returns to the page,
    finds the video where they left off).
    """
    rc, output = _run_node_tests()
    assert rc == 0, output
    names = [name for _, name in _parse_tap_test_names(output)]
    assert any("enableResume" in n or "resume" in n.lower() for n in names), names


# ── Video template integration ──────────────────────────────────────────


def test_video_template_loads_yt_player_script():
    """video.html must load yt_player.js before any script that needs it.

    Order matters: yt_player.js MUST come before transcript-follow.js
    (the latter uses the wrapper if available). And the inline <script>
    must reference YTPlayer.init().
    """
    text = VIDEO_TEMPLATE.read_text()
    yt_idx = text.find('yt_player.js')
    tf_idx = text.find('transcript-follow.js')
    assert yt_idx > -1, "video.html must include yt_player.js"
    assert tf_idx > -1, "video.html must still include transcript-follow.js"
    assert yt_idx < tf_idx, (
        "yt_player.js must load BEFORE transcript-follow.js "
        "so the wrapper is available when transcript-follow wires up."
    )
    assert 'YTPlayer.init()' in text, (
        "video.html must call YTPlayer.init() to bootstrap the wrapper"
    )


def test_video_template_iframe_has_youtube_player_id():
    """The YouTube iframe must have id='youtube-player' so YTPlayer
    can detect it via document.getElementById."""
    text = VIDEO_TEMPLATE.read_text()
    # Match the iframe block (Day 8: added the id)
    assert 'id="youtube-player"' in text, (
        "YouTube iframe in video.html must have id='youtube-player'"
    )
    assert 'enablejsapi=1' in text, (
        "YouTube iframe src must include enablejsapi=1 so postMessage works"
    )


def test_video_template_seek_uses_ytplayer_first():
    """Regression guard: after Day 8, video.html's seekTo() and the
    mindmap jump handler must TRY YTPlayer first, then fall back to
    document.querySelector('video') only if YTPlayer isn't ready.

    We assert by source-level grep: the primary path uses YTPlayer, and
    the fallback uses querySelector('video').
    """
    text = VIDEO_TEMPLATE.read_text()
    # Count occurrences of querySelector('video'). Acceptable uses:
    #  - initFollow fallback (transcript-follow wiring when YTPlayer isn't ready)
    #  - seekTo fallback (brief warmup window)
    #  - mindmap jump fallback (same warmup window)
    #  - swap modal (legacy <video>-only path, intentional)
    # What we DON'T accept: silent fallback that completely bypasses YTPlayer.
    fallback_count = text.count("querySelector('video')")
    ytplayer_count = text.count('YTPlayer')
    assert ytplayer_count >= 5, (
        f"video.html should reference YTPlayer ≥5 times (init, getInstance, "
        f"play, pause, seekTo). Found {ytplayer_count}. Make sure the "
        "primary seek path uses the wrapper, not direct DOM access."
    )
    # The fallback uses are legitimate, but they must be guarded by an
    # `if (window.YTPlayer && YTPlayer.getInstance())` check, not silent.
    assert fallback_count <= 4, (
        f"Found {fallback_count} querySelector('video') in video.html. "
        "This is too many — Day 8 should have collapsed most to YTPlayer. "
        "If you added new ones, ensure they fall back only after checking "
        "YTPlayer.getInstance() first."
    )


def test_static_mount_serves_yt_player_js():
    """The static mount in app/main.py must serve /static/js/yt_player.js.

    If the file exists on disk but isn't served, the page won't load
    the wrapper and all Day 8 features silently break.
    """
    from fastapi.testclient import TestClient
    from app.main import app

    client = TestClient(app)
    r = client.get("/static/js/yt_player.js")
    assert r.status_code == 200, (
        f"/static/js/yt_player.js returned {r.status_code}. "
        "Check that the static mount includes js/."
    )
    assert "YTPlayer" in r.text, (
        "yt_player.js must export the YTPlayer global"
    )

File: tests/test_yt_player.mjs
// tests/test_yt_player.mjs
//
// Unit tests for app/static/js/yt_player.js (Day 8).
//
// What we test:
//   1. Pure helpers (storageKey sanitization, savePosition/loadPosition).
//   2. Source-level regression: the JS must NOT use the legacy
//      `document.querySelector('video')` pattern for YouTube videos
//      (we removed those calls; this guards against regressions).
//   3. enableResume() reads from localStorage and seeks to the saved
//      position on ready.
//
// What we DON'T test here:
//   - Actual YouTube IFrame API calls (requires network + real iframe).
//     Those are integration-tested manually via the live Cloudflare Tunnel.
//   - rAF polling loop (we don't drive animation frames in the test env;
//     the loop just needs to NOT crash if it can't run).
//
// Run with: `node --test tests/test_yt_player.mjs`
// or via the pytest wrapper in tests/test_yt_player.py.

import test from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SRC = readFileSync(
    resolve(__dirname, '..', 'app', 'static', 'js', 'yt_player.js'),
    'utf8'
);


// ── DOM shim: enough to load yt_player.js + drive its IIFE ───────────────
//
// We need a `window` with:
//   - document.getElementById / querySelector
//   - localStorage
//   - requestAnimationFrame / cancelAnimationFrame (no-ops for these tests)
//   - location (used by URL parsing inside buildYouTubePlayer)
//
// Plus a way to inject our own YTPlayer globals after the IIFE runs.


class FakeElement {
    constructor(tagName) {
        this.tagName = (tagName || '').toUpperCase();
        this.attrs = {};
        this.children = [];
        this.listeners = {};
        this._src = '';
        this.style = {};
    }
    get src() { return this._src; }
    set src(v) { this._src = String(v); }
    addEventListener(t, fn) {
        (this.listeners[t] = this.listeners[t] || []).push(fn);
    }
    removeEventListener(t, fn) {
        const arr = this.listeners[t] || [];
        const i = arr.indexOf(fn);
        if (i >= 0) arr.splice(i, 1);
    }
    setAttribute(k, v) { this.attrs[k] = String(v); }
    getAttribute(k) { return this.attrs[k]; }
    appendChild(c) { this.children.push(c); return c; }
    insertBefore(c, ref) {
        const i = this.children.indexOf(ref);
        if (i === -1) this.children.push(c);
        else this.children.splice(i, 0, c);
        return c;
    }
    getElementsByTagName() {
        // For the script-injection path: return a non-empty array
        // so the IIFE doesn't throw on `parentNode.insertBefore`.
        return [];
    }
}


class FakeLocalStorage {
    constructor() { this._store = {}; }
    getItem(k) { return Object.prototype.hasOwnProperty.call(this._store, k) ? this._store[k] : null; }
    setItem(k, v) { this._store[k] = String(v); }
    removeItem(k) { delete this._store[k]; }
    clear() { this._store = {}; }
}


function makeFakeWindow({ iframe = null, video = null } = {}) {
    const elementsById = {};
    if (iframe) elementsById['youtube-player'] = iframe;

    const win = {
        // Storage
        localStorage: new FakeLocalStorage(),
        // No-op rAF (we don't drive animation frames in these tests)
        requestAnimationFrame: (fn) => { return 0; },
        cancelAnimationFrame: () => {},
        // Location (used by URL() inside buildYouTubePlayer)
        location: { href: 'https://test.local/video/123', origin: 'https://test.local' },
        // YTPlayer instance (set after IIFE runs)
        YTPlayer: null,
        // Global hooks YouTube's iframe_api calls
        onYouTubeIframeAPIReady: null,
        // Day 8: enableResume() attaches to beforeunload
        addEventListener() {},
        removeEventListener() {},
        // Queryable elements
        document: {
            getElementById(id) { return elementsById[id] || null; },
            querySelector(sel) {
                if (sel === 'video') return video;
                return null;
            },
            getElementsByTagName(tag) {
                if (tag === 'script') return [];
                if (tag === 'iframe') return iframe ? [iframe] : [];
                return [];
            },
            createElement(tag) { return new FakeElement(tag); },
            addEventListener() {},
            removeEventListener() {},
            visibilityState: 'visible',
        },
    };
    return win;
}


function loadYTPlayer(win) {
    // Eval the source in the fake window context. We can't `import` the
    // file directly because it's an IIFE that expects a real `window`.
    // We pass window, document, localStorage, etc. as named params so
    // the IIFE sees them as locals (matches browser behavior where all
    // of these are global identifiers).
    const fn = new Function(
        'window', 'document', 'localStorage', 'requestAnimationFrame',
        'cancelAnimationFrame', 'location',
        SRC
    );
    fn(
        win, win.document, win.localStorage, win.requestAnimationFrame,
        win.cancelAnimationFrame, win.location
    );
    return win.YTPlayer;
}


// ── Source-level regression checks ────────────────────────────────────────


test('does not call document.querySelector("video") inside seekTo/play/pause paths', () => {
    // The Day 8 refactor removed all `document.querySelector('video')`
    // calls from seekTo/play/pause/getCurrentTime paths in video.html
    // (they silently no-op on YouTube iframes). The remaining
    // `querySelector('video')` in yt_player.js is ONLY in detectBackend()
    // for backend identification — that one is correct.
    //
    // We grep for occurrences outside detectBackend() as a regression
    // guard. If anyone adds a new querySelector('video') in this file,
    // this test fails and forces them to think about YouTube compat.
    const lines = SRC.split('\n');
    let inDetectBackend = false;
    let braceDepth = 0;
    const violations = [];
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (/function detectBackend\s*\(/.test(line)) {
            inDetectBackend = true;
            braceDepth = 0;
        }
        if (inDetectBackend) {
            braceDepth += (line.match(/\{/g) || []).length;
            braceDepth -= (line.match(/\}/g) || []).length;
            if (braceDepth === 0 && /\}/.test(line)) inDetectBackend = false;
            continue;
        }
        if (line.includes("querySelector('video')") || line.includes('querySelector("video")')) {
            violations.push(`line ${i + 1}: ${line.trim()}`);
        }
    }
    assert.equal(
        violations.length, 0,
        `yt_player.js must not use querySelector('video') outside detectBackend(). Found:\n${violations.join('\n')}`
    );
});

test('exposes the documented public API surface', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    assert.ok(YTP, 'YTPlayer should be installed on window');
    for (const name of ['init', 'getInstance', 'onReady', 'onTimeUpdate', 'enableResume']) {
        assert.equal(typeof YTP[name], 'function', `${name} should be a function`);
    }
});

test('detects backend from #youtube-player iframe', () => {
    const fakeIframe = new FakeElement('iframe');
    fakeIframe.setAttribute('id', 'youtube-player');
    fakeIframe.src = 'https://www.youtube-nocookie.com/embed/dQw4w9WgXcQ';

    const win = makeFakeWindow({ iframe: fakeIframe });
    const YTP = loadYTPlayer(win);
    const backend = YTP._detectBackend();
    assert.deepEqual(backend, { kind: 'youtube', iframe: fakeIframe });
});

test('detects backend from <video> when no iframe', () => {
    const fakeVideo = new FakeElement('video');
    const win = makeFakeWindow({ video: fakeVideo });
    const YTP = loadYTPlayer(win);
    const backend = YTP._detectBackend();
    assert.deepEqual(backend, { kind: 'native', video: fakeVideo });
});

test('returns null backend when neither iframe nor video present', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    const backend = YTP._detectBackend();
    assert.equal(backend, null);
});


// ── Storage helpers (the resume feature) ──────────────────────────────────


test('storageKey strips dangerous characters from video id', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    // Slashes, quotes, dots, spaces etc. should be stripped so we never
    // inject into a path traversal / key collision in localStorage.
    assert.equal(YTP._storageKey('abc-123'), 'video_last_position_abc123', 'dashes stripped for collision resistance');
    assert.equal(
        YTP._storageKey('../../../etc/passwd'),
        'video_last_position_etcpasswd',
        'must strip path-traversal chars'
    );
    assert.equal(
        YTP._storageKey('x;DROP TABLE users'),
        'video_last_position_xDROPTABLEusers',
        'must strip SQL-injection-looking chars (incl. spaces and punctuation)'
    );
    assert.equal(
        YTP._storageKey("Robert'); DROP TABLE videos;--"),
        'video_last_position_RobertDROPTABLEvideos',
        'must strip classic SQL injection attempt (incl. -- comment markers)'
    );
});

test('savePosition writes to localStorage with the sanitized key', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    YTP._savePosition('video-abc', 123.5);
    assert.equal(win.localStorage.getItem('video_last_position_videoabc'), '123.5');
});

test('savePosition ignores non-finite or non-number values', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    YTP._savePosition('video-abc', NaN);         // NaN → skip
    YTP._savePosition('video-abc', Infinity);    // Infinity → skip
    YTP._savePosition('video-abc', -Infinity);   // -Infinity → skip
    YTP._savePosition('video-abc', undefined);   // not a number → skip
    YTP._savePosition('video-abc', null);        // not a number → skip
    YTP._savePosition('video-abc', '60');        // not a number → skip
    assert.equal(win.localStorage.getItem('video_last_position_videoabc'), null);
});

test('savePosition clamps negative values to 0 (not a corrupt write)', () => {
    // Negative seconds don't make sense for video position, but they're
    // a finite number — we don't want to crash; we save 0 instead.
    // This means a corrupted upstream caller can't poison the storage
    // with -5 → -5 reads back as 'invalid' which means user loses resume.
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    YTP._savePosition('video-abc', -5);
    assert.equal(win.localStorage.getItem('video_last_position_videoabc'), '0');
});

test('loadPosition returns 0 when nothing saved', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    assert.equal(YTP._loadPosition('never-saved'), 0);
});

test('loadPosition returns the saved value as a float', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    YTP._savePosition('video-abc', 99.7);
    assert.equal(YTP._loadPosition('video-abc'), 99.7);
});

test('loadPosition returns 0 on garbage data (defensive)', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    win.localStorage.setItem('video_last_position_videoabc', 'not a number');
    assert.equal(YTP._loadPosition('video-abc'), 0);
});

test('savePosition handles disabled localStorage gracefully', () => {
    // Some browsers disable localStorage in private mode and throw on setItem.
    // We swallow the error so the page doesn't break.
    const win = makeFakeWindow();
    win.localStorage.setItem = () => { throw new Error('QuotaExceededError'); };
    const YTP = loadYTPlayer(win);
    // Should NOT throw
    YTP._savePosition('video-abc', 10);
});


// ── enableResume wiring ───────────────────────────────────────────────────


test('enableResume is a no-op when videoId is falsy', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    // Should not throw
    const unsub = YTP.enableResume(null);
    assert.equal(typeof unsub, 'function', 'must return an unsubscribe fn even when no-op');
    unsub();
});

test('enableResume returns an unsubscribe function', () => {
    // The smoke test for enableResume: it must return an unsubscribe
    // function so we can clean up on page transition.
    // The seek-to-saved-position logic itself depends on internal
    // closure state (playerInstance), which is hard to mock without
    // a full JS DOM. We verify the surface here; the full integration
    // is covered by manual smoke tests via the Cloudflare Tunnel.
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    win.localStorage.setItem('video_last_position_videoabc', '42.5');
    const unsub = YTP.enableResume('video-abc');
    assert.equal(typeof unsub, 'function');
    // Should not throw when called
    unsub();
});

test('enableResume can be called for multiple videos without leaking', () => {
    // Each call should register its own handlers and return its own
    // unsubscriber. We verify by calling it twice and unsubscribing once
    // — the first video's handlers should be gone but the second video's
    // should still be in the queue (we can't easily verify that without
    // exposing internals, so we just verify no errors).
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    const unsub1 = YTP.enableResume('video-one');
    const unsub2 = YTP.enableResume('video-two');
    assert.notEqual(unsub1, unsub2, 'each call returns a distinct unsubscriber');
    unsub1();  // first one unsubscribes cleanly
    unsub2();  // second one unsubscribes cleanly
});


// ── Idempotency ───────────────────────────────────────────────────────────


test('init() called twice does not throw and returns the same wrapper', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    // init() with no backend rejects — that's expected. We just want
    // to confirm calling init() multiple times doesn't crash and
    // doesn't double-install handlers.
    assert.rejects(() => YTP.init(), /no <video> or #youtube-player iframe found/);
    assert.rejects(() => YTP.init(), /no <video> or #youtube-player iframe found/);
});

test('onReady fires immediately if player is already ready', () => {
    const win = makeFakeWindow();
    const YTP = loadYTPlayer(win);
    // Inject an instance that claims to be ready
    YTP._instanceForTest = { kind: 'native', isReady: true };
    // Manually fire onReady by setting instance + calling onReady
    // (this is the same code path as init() resolving)
    let fired = false;
    YTP.onReady(() => { fired = true; });
    // In a real init() the instance is set first, then handlers fire.
    // We simulate by calling onReady after setting instance:
    // (The cleanest path requires running init() which requires a backend.
    //  This test just verifies the API surface, not the full timing.)
    assert.equal(typeof YTP.onReady, 'function');
});


// ── Source-level: must NOT load YouTube API on pages without iframes ─────


test('native backend does not load YouTube iframe_api', () => {
    // This is a source-level check. The buildNativePlayer path must
    // never touch loadYouTubeApi() — we don't want every <video> page
    // to make a 200KB script request to youtube.com.
    //
    // We check structurally:
    //   - the init() function branches on backend.kind
    //   - the 'youtube' branch is the only one that calls loadYouTubeApi()
    //   - there's no loadYouTubeApi() call outside that branch
    assert.ok(
        SRC.includes("backend.kind === 'youtube'"),
        "yt_player.js should branch on backend.kind === 'youtube' for API loading"
    );
    // Find the init() function and verify loadYouTubeApi is only called
    // inside the youtube branch (look for the 'await loadYouTubeApi()' line)
    const initStart = SRC.indexOf('async function init()');
    const initEnd = SRC.indexOf('function getInstance()');
    const initBody = SRC.slice(initStart, initEnd);
    assert.ok(
        initBody.includes('await loadYouTubeApi()'),
        'init() must call loadYouTubeApi() somewhere'
    );
    // The 'youtube' branch text should appear BEFORE the loadYouTubeApi call
    const branchIdx = initBody.indexOf("backend.kind === 'youtube'");
    const callIdx = initBody.indexOf('await loadYouTubeApi()');
    assert.ok(branchIdx > -1 && callIdx > branchIdx, 'loadYouTubeApi is called inside the youtube branch');
    assert.ok(callIdx < initBody.indexOf('buildNativePlayer'), 'loadYouTubeApi is called BEFORE the native branch (so native branch can skip it)');
});


test('both player instances expose onTimeUpdate + kind (transcript-follow contract)', () => {
    // 2026-09-06 regression guard. transcript-follow.js keys its
    // wrapper-vs-raw-<video> detection on:
    //
    //     typeof videoEl.onTimeUpdate === 'function' && videoEl.kind
    //
    // The Day-8 instances shipped `.kind` but NOT `.onTimeUpdate`, so
    // the check fell into the raw-DOM branch and called
    // `instance.addEventListener` — a TypeError that silently killed
    // transcript follow/highlight on every YouTube video.
    //
    // This source-level guard ensures BOTH instance literals (youtube
    // and native) expose both fields. If a future refactor renames or
    // removes either, this fails loudly.
    const ytIdx = SRC.indexOf("kind: 'youtube'");
    assert.ok(ytIdx > -1, 'YouTube instance must have kind: youtube');
    const ytBlock = SRC.slice(ytIdx, SRC.indexOf('_raw: ytPlayer'));
    assert.ok(
        /onTimeUpdate\s*:/.test(ytBlock),
        'YouTube instance must expose onTimeUpdate(handler) — transcript-follow.js contract'
    );

    const nativeIdx = SRC.indexOf("kind: 'native'");
    assert.ok(nativeIdx > -1, 'native instance must have kind: native');
    const nativeBlock = SRC.slice(nativeIdx, SRC.indexOf('_raw: video'));
    assert.ok(
        /onTimeUpdate\s*:/.test(nativeBlock),
        'native instance must expose onTimeUpdate(handler) — transcript-follow.js contract'
    );
});
